File: python/01-arrays-hashing/character_count.py
"""
Problem: Character Count / Run-Length Encoding
Pattern: Arrays & Hashing (string traversal)
Time: O(n) | Space: O(n)

count_char_serialize: run-length-encode a string -- each run of repeated
characters becomes "<char><count>".
count_all_char: count total occurrences of every character in a string.

Input:  count_char_serialize("abbbccddddaab")
Output: "a1b3c2d4a2b1"

Input:  count_all_char("aabbbc")
Output: {'a': 2, 'b': 3, 'c': 1}
"""
import re
from collections import Counter

def count_char_serialize(s):
    count = 1
    result = ''
    for i in range(len(s)):
        if i + 1 < len(s) and s[i] == s[i + 1]:
            count += 1
        else:
            result += s[i] + str(count)
            count = 1
    return result


def count_all_char(s):
    cleaned = re.sub(r'[^a-z0-9]', '', s, flags=re.IGNORECASE)
    return dict(Counter(cleaned))


if __name__ == "__main__":
    print(count_char_serialize("abbbccddddaab"))  # a1b3c2d4a2b1
    print(count_all_char("aabbbc"))                # {'a': 2, 'b': 3, 'c': 1}

File: javascript/01-arrays-hashing/characterCount.js

/**
 * Problem: Character Count / Run-Length Encoding
 * Pattern: Arrays & Hashing (string traversal)
 * Time: O(n) | Space: O(n)
 *
 * countCharSerlize: run-length-encode a string -- each run of repeated
 * characters becomes "<char><count>".
 * countAllChar: count total occurrences of every character in a string
 * (not run-based -- occurrences anywhere, not just consecutive).
 *
 * Input:  countCharSerlize("abbbccddddaab")
 * Output: "a1b3c2d4a2b1"
 *
 * Input:  countAllChar("aabbbc")
 * Output: { a: 2, b: 3, c: 1 }
 */
function countCharSerlize(string) {
    let str = string;
    let count = 1;
    let result = '';
     for(let i = 0; i <  str.length ;i++) {
         if(str[i] == str[i+1]) {
             count ++
         } else {
            result += str[i] + count;
            count = 1;
         }
     }
    return result;
}

function countAllChar(string) {
   const arrString = string.replace(/[^a-z0-9]/gi,'').split('');
   return arrString.reduce((count, item) =>{
      count[item] = (count[item] || 0) + 1;
       return count;  
   } ,{}); 
}

module.exports = { countCharSerlize, countAllChar};